presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorNodePartitioningProvider.java

```diff
@@ -42,7 +43,12 @@
         return singletonList(NOT_PARTITIONED);
     }
 
-    Map<Integer, Node> getBucketToNode(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle);
+    int getBucketCount(ConnectorPartitioningHandle partitioningHandle);
+
+    default Optional<Map<Integer, Node>> getBucketToNode(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
```

Comment:
I think we can add `hasBucketToNodeMapping` method.

Reply:
Another alternative would be to merge `getBucketCount` and `getBucketToNode`, and have the new method return a struct. The struct should have two constructors: one that takes an int, and one that takes a map.  I don't like the current API because `getBucketCount` is redundant when `getBucketToNode` does not return `empty`. Engine could also misbehave if the two method mismatch, making it necessary to have validation on engine side. However, the validation code would be duplicate of the code that generated the integer in the first place.